### PR TITLE
docs: fix remaining uses of "by reference" for calling convention

### LINKED ISSUE #23127: Documentation: Misleading statement about passing behavior of immutables
The following statement seems wrong ('Documentation / Type / Mutable Composite Types'):

> * An object with an immutable type is passed around (both in assignment statements and in function calls) by copying, whereas a mutable type is passed around by reference.

This reads to me as if values are copied, which seems not to be the case. See 'Documentation / Functions / Argument Passing Behavior':

> Julia function arguments follow a convention sometimes called "pass-by-sharing", which means that values are not copied when they are passed to functions. Function arguments themselves act as new variable bindings (new locations that can refer to values), but the values they refer to are identical to the passed values.

Edit: removed misleading example

### PR DESCRIPTION
**Pull request by StefanKarpinski**
All uses of "by reference" in docs are technically incorrect. Several of the explanations of behavior around this were also wrong. This corrects them. Inspired by https://github.com/JuliaLang/julia/pull/26427.

closes #23127

### CODE REVIEW

**Comment by giordano** [comment]
Perhaps you could add a link to https://en.m.wikipedia.org/wiki/Evaluation_strategy#Call_by_sharing? I think call by sharing is less known compared to the other strategies.

**Comment by StefanKarpinski** [comment]
Good idea. Updated.

**Comment by jrklasen** [comment]
closes #23127

**Review by ararslan — APPROVED** [review]

File: doc/src/devdocs/callconv.md [hunk 1]
```diff
@@ -16,7 +16,8 @@ signature.
 
   * LLVM ghosts (zero-length types) are omitted.
   * LLVM scalars and vectors are passed by value.
-  * LLVM aggregates (arrays and structs) are passed by reference.
+  * LLVM aggregates (arrays and structs) are [passed by
+    sharing](https://en.wikipedia.org/wiki/Evaluation_strategy#Call_by_sharing) (i.e. as pointers).
```
Comment (JeffBezanson) [inline doc/src/devdocs/callconv.md]: This refers to the ABI, so I don't think the call-by-sharing terminology is really necessary or helpful.

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 2]
```diff
@@ -26,11 +26,16 @@ A function name may be used alone in place of the tuple (just `:function` or `"f
...
-and so to call a Fortran function via [`ccall`](@ref) you must pass the mangled identifier corresponding
-to the rule followed by your Fortran compiler.  Also, when calling a Fortran function, all inputs
-must be passed by reference.
+By default, Fortran compilers [generate mangled
+names](https://en.wikipedia.org/wiki/Name_mangling#Fortran) (for example,
+converting function names to lowercase or uppercase, often appending an
+underscore), and so to call a Fortran function via [`ccall`](@ref) you must pass
+the mangled identifier corresponding to the rule followed by your Fortran
+compiler.  Also, when calling a Fortran function, all inputs must be passed as
+pointers to heap-allocated values. This applies not only to arrays and other
```
Comment (JeffBezanson) [inline doc/src/manual/calling-c-and-fortran-code.md]: They don't have to be heap-allocated.

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 3]
```diff
@@ -763,11 +770,13 @@ that has no internal fields and whose sole purpose is to be placed in the type p
...
-The input `n` is passed by value, and so the function's input signature is simply declared as
-`(Csize_t,)` without any `Ref` or `Ptr` necessary. (If the wrapper was calling a Fortran function
-instead, the corresponding function input signature should instead be `(Ref{Csize_t},)`, since
-Fortran variables are passed by reference.) Furthermore, `n` can be any type that is convertable
-to a `Csize_t` integer; the [`ccall`](@ref) implicitly calls [`Base.cconvert(Csize_t, n)`](@ref).
+The input `n` is passed by value, and so the function's input signature is
+simply declared as `(Csize_t,)` without any `Ref` or `Ptr` necessary. (If the
+wrapper was calling a Fortran function instead, the corresponding function input
+signature should instead be `(Ref{Csize_t},)`, since Fortran variables are
+passed by pointers.) Furthermore, `n` can be any type that is convertable to a
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]: This should still say "by reference", since we're referring to Fortran here.

"GNU Fortran passes most arguments by reference, i.e. by passing a pointer to the data" (https://gcc.gnu.org/onlinedocs/gfortran/Argument-passing-conventions.html)

File: doc/src/manual/arrays.md [hunk 4]
```diff
@@ -17,11 +17,17 @@ be written in a vectorized style for performance. Julia's compiler uses type inf
...
-arrays within a function will be visible in the parent function. The entire Julia array library
-ensures that inputs are not modified by library functions. User code, if it needs to exhibit similar
-behavior, should take care to create a copy of inputs that it may modify.
+In Julia, all arguments to functions are [passed by
+sharing](https://en.wikipedia.org/wiki/Evaluation_strategy#Call_by_sharing)
+(i.e. by pointers). Some technical computing languages pass arrays by value, and
+while this prevents accidental modification by callees of a value in the caller,
+it makes avoiding unwanted copying of arrays difficult. In Julia, modifications
+made to input arrays within a function will be visible in the parent function.
+Unless a function ends with a `!` indicating that it mutates one or more of its
```
Comment (vtjnash) [inline doc/src/manual/arrays.md]: Alternate wording? "By convention, a function name ending with a `!` indicates that it will mutate or destroy the value of one or more of its arguments. Callees must make explicit copies to ensure that they don't modify inputs that they don't intent to change. Many non-mutating functions are implemented by calling a function of the same name with an added `!` at the end on an explicit copy of the input, and returning that copy."

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 3 above)
Comment (StefanKarpinski) [reply doc/src/manual/calling-c-and-fortran-code.md]: It should not since there's a different and conflicting meaning of the term even though Fortran did use it in a different way first.

Comment (vtjnash) [reply doc/src/manual/calling-c-and-fortran-code.md]: IIRC, Fortran actually does pass arguments by reference though, since it doesn't have pointers and the variables themselves are values, not bindings.

Comment (StefanKarpinski) [reply doc/src/manual/calling-c-and-fortran-code.md]: Yes, but nothing from Julia's perspective is "passed by reference" – from Julia's perspective, mutable objects are passed by sharing to a function which cannot change any bindings in the caller but can mutate the shared objects.

Comment (vtjnash) [reply doc/src/manual/calling-c-and-fortran-code.md]: Yes, that's fine. But the text above is literally "... Fortran variables are passed by reference". We can reword it to clarify that the subject is Julia "..., since Fortran expects Julia to pass a pointer to the value." But in the above clause, the subject is Fortran. Another way to say it is that "..., since Fortran passes variables by reference, which is equivalent in Julia to passing a pointer"

Comment (StefanKarpinski) [reply doc/src/manual/calling-c-and-fortran-code.md]: That would be fine but saying that they are "passed by pointers" is also correct.
